src/rest-server/src/templates/yarnContainerScript.mustache

```diff
@@ -26,44 +26,230 @@ export PAI_DEFAULT_FS_URI={{ hdfsUri }}
+  echo "solution: $solution" >> $RUNTIME_AGG_ERROR_FILE
+  echo "originalUserExitCode: $user_exit_code" >> $RUNTIME_AGG_ERROR_FILE
+
+  echo "dockerStatus: $docker_status" >> $RUNTIME_AGG_ERROR_FILE
+  echo "caughtSignal: $caught_signal" >> $RUNTIME_AGG_ERROR_FILE
+
+  # Not supported yet
+  echo "caughtException:" >> $RUNTIME_AGG_ERROR_FILE
+  echo "matchedUserLogString:" >> $RUNTIME_AGG_ERROR_FILE
+}
+
+
+function printf_file_tail_with_shift()
+{
+  local file_name=$1
+  local shift_space=$2
+  local tail_num=$3
+
+  if [[ $tail_num == "" ]]; then
+    tail_num=10
+  fi
+
+  local file_content=$(tail -n "$tail_num" "$file_name")
+  while IFS= read -r line
+  do
+    line=$(echo $line | sed -e 's/^[[:blank:]]*//' | tr -d '\n\r')
+    echo "$shift_space$line"
+  done < <(printf '%s\n' "$file_content")
+}
+
+
+function generate_error_summary()
+{
+  if [[ -f $DISK_CLEANER_ERROR_FILE ]]; then
+    grep "ACTION \"KILL\"" $DISK_CLEANER_ERROR_FILE 2>&1 > /dev/null
+    if [[ $? -eq 0 ]]; then
+      write_error_summary $CONTAINER_OOD_ERROR "diskCleanerErrorLog" $DISK_CLEANER_ERROR_FILE
+      return
+    fi
+  fi
+
+  if [[ -f $RUNTIME_DOCKER_ERR_LOG ]]; then
+    grep "ERROR REASON" $RUNTIME_DOCKER_ERR_LOG | grep "Exiting because of user error" 2>&1 > /dev/null
+    if [[ $? -eq 0 ]]; then
+      write_error_summary $RUNTIME_UNRECOGNIZED_ERROR "originalUserExitCode" $RUNTIME_DOCKER_ERR_LOG
+      return
+    fi
+  fi
+
+  if [[ -f $RUNTIME_YARN_ERR_LOG ]]; then
+    grep "ERROR REASON" $RUNTIME_YARN_ERR_LOG | grep "docker deamon error" 2>&1 > /dev/null
+    if [[ $? -eq 0 ]]; then
+      write_error_summary $DOCKER_DEAMON_ERROR "dockerStatus" $RUNTIME_YARN_ERR_LOG "DeamonError"
+      return
+    fi
+
+    grep "ERROR REASON" $RUNTIME_YARN_ERR_LOG | grep "out of memory (OOM)" 2>&1 > /dev/null
+    if [[ $? -eq 0 ]]; then
+      local signal=$(grep -o "RECEIVE_SIGNAL.*" $RUNTIME_YARN_ERR_LOG | grep -o '".*"' | sed 's/"//g')
+      if [[ $signal == "" ]]; then
+        signal="unknownSignal"
+      fi
+      write_error_summary $CONTAINER_OOM_ERROR "caughtSignal" $RUNTIME_YARN_ERR_LOG "" $signal
+      return
+    fi
+
+    grep "ERROR REASON" $RUNTIME_YARN_ERR_LOG | grep "YARN Container receive signal SIGTERM" 2>&1 > /dev/null
+    if [[ $? -eq 0 ]]; then
+      local signal=$(grep -o "RECEIVE_SIGNAL.*" $RUNTIME_YARN_ERR_LOG | grep -o '".*"' | sed 's/"//g')
+      if [[ $signal == "" ]]; then
+        signal="unknownSignal"
+      fi
+      write_error_summary $CONTAINER_SIGTERM_ERROR "caughtSignal" $RUNTIME_YARN_ERR_LOG "" $signal
+    fi
+  fi
+
+  # write empty error summary for successful exit
+  write_error_summary
+}
+
+
+function generate_agg_err()
+{
+  # Here we persume the kill event trigger seq is diskCleaner, runtime and user error
+  printf "[DEBUG] Generate aggregate dynamic error info\n"
+
+  printf "[PAI_RUNTIME_ERROR_START]\n" >> $RUNTIME_AGG_ERROR_FILE
+  generate_error_summary
+
+  echo "errorLogs:" >> $RUNTIME_AGG_ERROR_FILE
```

Comment:
Have you Limited the size of Runtime.pai.agg.error file < 4KB

Reply:
I limit the retrieved log lines no more than 35 lines. If you need the size strict no more than 4K, I can change the tail command to only retrieve the limited number of characters. But it may cause some lines be cut.

Reply:
strict no more than 4K, log line cut is fine

Reply:
fixed